src/Sylius/Bundle/CoreBundle/Fixture/Factory/PromotionExampleFactory.php

```diff
@@ -142,6 +153,32 @@ protected function configureOptions(OptionsResolver $resolver): void
 
                 return $actions;
             })
+
+            ->setDefault('coupons', [])
+            ->setAllowedTypes('coupons', 'array')
+            ->setNormalizer('coupons', function (Options $options, array $couponDefinitions): array {
+                if (!$options['coupon_based'] && count($couponDefinitions) > 0) {
+                    throw new InvalidArgumentException('Cannot define coupons for promotion that is not coupon based');
+                }
+
+                $coupons = [];
+                foreach ($couponDefinitions as $couponDefinition) {
+                    /** @var PromotionCouponInterface $coupon */
+                    $coupon = $this->couponFactory->createNew();
+                    $coupon->setCode($couponDefinition['code']);
+                    $coupon->setPerCustomerUsageLimit($couponDefinition['per_customer_usage_limit']);
+                    $coupon->setReusableFromCancelledOrders($couponDefinition['reusable_from_cancelled_orders']);
+                    $coupon->setUsageLimit($couponDefinition['usage_limit']);
+
+                    if (null !== $couponDefinition['expires_at']) {
+                        $coupon->setExpiresAt(new \DateTime($couponDefinition['expires_at']));
+                    }
+
+                    $coupons[] = $coupon;
+                }
+
+                return $coupons;
```

Comment:
We could also extract it to the private/protected method for better readability 🖖